src/analysis/journey_events_analysis.py

```diff
@@ -0,0 +1,239 @@
+import argparse
+import logging.config
+import os
+from ast import literal_eval
+from collections import Counter
+
+import pandas as pd
+from scipy import stats
+
+AGGREGATE_COLUMNS = ['DeviceCategories', 'Event_cats_agg', 'Event_cat_act_agg']
+
+NAVIGATE_EVENT_CATS = ['breadcrumbClicked',
+                       'homeLinkClicked',
+                       '/search',
+                       'navDocumentCollectionLinkClicked',
+                       'navAccordionLinkClicked',
+                       'navLeafLinkClicked',
+                       'navPolicyAreaLinkClicked',
+                       'navServicesInformationLinkClicked',
+                       'navSubtopicContentItemLinkClicked',
+                       'navSubtopicLinkClicked',
+                       'navTopicLinkClicked',
+                       'relatedTaxonomyLinkClicked',
+                       'stepNavHeaderClicked', 'stepNavLinkClicked', 'stepNavPartOfClicked']
+
+# Useful for explicit event category and action matching, may extend in the future
+NAVIGATE_EVENT_CATS_ACTS = [('relatedLinkClicked', 'Explore the topic')]
+
+
+def device_count(x, device):
+    return sum([value for item, value in x if item == device])
+
+
+def has_related_event(sequence_str):
+    return all(cond in sequence_str for cond in ["relatedLinkClicked", "Related content"])
+
+
+def has_nav_event_cat(sequence_str):
+    return any(event_cat in sequence_str for event_cat in NAVIGATE_EVENT_CATS)
+
+
+def has_nav_event_cat_act(sequence_str):
+    return any(
+        event_cat in sequence_str and event_act in sequence_str for event_cat, event_act in NAVIGATE_EVENT_CATS_ACTS)
+
+
+def map_device_counter(df):
+    """
+    Count the device-based occurrences per target device
+    :param df:
+    :return:
+    """
+    logging.info("Mapping device counts")
+    df["DesktopCount"] = df['DeviceCategories'].map(lambda x: device_count(x, "desktop"))
+    df["MobileCount"] = df['DeviceCategories'].map(lambda x: device_count(x, "mobile"))
```

Comment:
Worth commenting that Tablet is assumed to have been filtered out in BQ step?

Reply:
Maybe, tbh from now on this will be computed within BQ